File: schemas/auction/drizzle/bid_increment_rules.ts
// bid_increment_rules: Defines price-range-based bid increment tiers for auctions.
// See README.md for full design rationale.

import { pgTable, uuid, integer, timestamp, index } from "drizzle-orm/pg-core";
import { auctions } from "./auctions";

export const bidIncrementRules = pgTable(
  "bid_increment_rules",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    auctionId: uuid("auction_id").references(() => auctions.id, { onDelete: "cascade" }),
    minPrice: integer("min_price").notNull(),
    maxPrice: integer("max_price"),
    increment: integer("increment").notNull(),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    index("idx_bid_increment_rules_auction_id").on(table.auctionId),
    index("idx_bid_increment_rules_min_price").on(table.minPrice),
  ]
);

File: schemas/auction/spacetimedb/bid_increment_rules.rs
// bid_increment_rules: Price-range rules that determine the minimum bid increment for an auction.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[spacetimedb::table(name = bid_increment_rules, public)]
pub struct BidIncrementRule {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub auction_id: Option<String>, // UUID. FK → auctions.id (cascade delete). Null = global default rule.

    #[index(btree)]
    pub min_price: f64, // Lower bound of the price range (inclusive).

    pub max_price: Option<f64>, // Upper bound of the price range. Null = no upper limit.

    pub increment: f64, // Minimum bid increment required within this price range.

    pub created_at: Timestamp,
}
